tests/test_fastapi_roles.py

```diff
@@ -0,0 +1,341 @@
+            try:
+                response = await call_next(request)
+                return response
+            finally:
+                DBOS.set_authentication(prev_user, prev_roles)
+
+    @app.get("/dboserror")
+    def test_dbos_error() -> None:
+        raise DBOSNotAuthorizedError("test")
+
+    @app.get("/dbosinternalerror")
+    def test_dbos_error_internal() -> None:
+        raise DBOSDuplicateWorkflowEventError("nosuchwf", "test")
+
+    @app.get("/open/{var1}")
+    @DBOS.required_roles([])
+    @DBOS.workflow()
+    def test_open_endpoint(var1: str) -> str:
+        result = var1
+        ctx = assert_current_dbos_context()
+        return result
+
+    @app.get("/user/{var1}")
+    @DBOS.required_roles(["user"])
+    @DBOS.workflow()
+    def test_user_endpoint(var1: str) -> str:
+        result = var1
+        ctx = assert_current_dbos_context()
+        assert ctx.assumed_role == "user"
+        return result
+
+    @app.get("/engineer/{var1}")
+    @DBOS.required_roles(["engineer"])
+    @DBOS.workflow()
+    def test_engineer_endpoint(var1: str) -> str:
+        result = var1
+        ctx = assert_current_dbos_context()
+        assert ctx.assumed_role == "engineer"
+        return result
+
+    @app.get("/admin/{var1}")
+    @DBOS.required_roles(["admin"])
+    @DBOS.workflow()
+    def test_admin_endpoint(var1: str) -> str:
+        result = var1
+        ctx = assert_current_dbos_context()
+        assert ctx.assumed_role == "admin"
+        return result
+
+    @DBOS.required_roles(["admin"])
+    @DBOS.workflow()
+    def test_admin_endpoint_nh(var1: str) -> str:
+        result = var1
+        ctx = assert_current_dbos_context()
+        assert ctx.assumed_role == "admin"
+        return result
+
+    @app.get("/adminalt/{var1}")
+    def test_admin_handler(var1: str) -> str:
+        return test_admin_endpoint_nh(var1)
+
+    @app.get("/error")
+    def test_error() -> None:
+        raise HTTPException(status_code=401)
+
+    client = TestClient(app)
+
+    response = client.get("/user/b")
+    assert response.status_code == 200
+    assert response.text == '"b"'
+
+    # Test for roles set into span
+    # Get the spans that were recorded
+    spans = exporter.get_finished_spans()
+
+    # Assert that we have two spans
+    #  One is the handler span
+    #  The other is the WF span
+    assert len(spans) == 2
+    assert spans[0].context is not None
+    assert spans[1].context is not None
+
+    span = spans[1]
+    assert span.name == "/user/b"
+    assert span.attributes is not None
+
+    span = spans[0]
+    assert span.name == "test_user_endpoint"
+    assert span.parent is not None
+    assert span.parent.span_id == spans[1].context.span_id
+    assert span.attributes is not None
+    assert span.attributes["authenticatedUser"] == "user1"
+    assert span.attributes["authenticatedUserAssumedRole"] == "user"
+    assert span.attributes["authenticatedUserRoles"] == '["user", "engineer"]'
+
```

Comment:
You may need to wait for system buffers to flush. Use `wait_for_buffer_flush` in system DB

Reply:
Actually, you only need to wait if that's a single-transaction temp workflow. Inputs and status are synchronously written for other types of workflows.